### PR TITLE
Add Fake IdP SAML App tokens: Backend and tests

### PR DESCRIPTION
**Pull request by wleightond**
## Proposed changes
This PR adds the backend and test components for fake IdP SAML2 app tokens. 

These are installed as a SAML app on the user's Identity Provider dashboard. When an attacker clicks on the app a SAML request is sent to the canarytokens server with the user's identity, and an alert is triggered.

Changes:
- Add IdP token trigger handling to backend
- Update the notification template
- Add unit and integration tests for IdP tokens
- Clean up models

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] Lint and unit tests pass locally with my changes (if applicable)
- [x] I have run pre-commit (`pre-commit` in the repo)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation in [canarytokens-docs PR 57](https://github.com/thinkst/canarytokens-docs/pull/57)

### CODE REVIEW

File: canarytokens/tokens.py [hunk 1]
```diff
@@ -598,6 +599,24 @@ def _get_response_for_fast_redirect(
...
+        return http_general_info, {"src_data": src_data}
+
+    @staticmethod
+    def _get_response_for_idp_app(canarydrop: canarydrop.Canarydrop, request: Request):
+        if canarydrop.redirect_url:
+            if canarydrop.browser_scanner_enabled:
+                return Canarytoken._get_response_for_slow_redirect(canarydrop, request)
+            return Canarytoken._get_response_for_fast_redirect(canarydrop, request)
+        return Canarytoken._get_response_for_web(canarydrop, request)
+
```
Comment (gjcthinkst) [inline canarytokens/tokens.py:619]:
```suggestion
        if not canarydrop.redirect_url:
            return Canarytoken._get_response_for_web(canarydrop, request)
        if not canarydrop.browser_scanner_enabled:
            return Canarytoken._get_response_for_fast_redirect(canarydrop, request)
        
        return Canarytoken._get_response_for_slow_redirect(canarydrop, request)
```

I think it makes it a bit easier to understand if we reduce the indentation.

File: tests/integration/test_saml_token.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        login_url, data, headers={"Accept": "text/html"}, timeout=(30, 30)
+    )
+    assert resp.status_code == OK
+
+    # Check that the returned history has a single hit
+    stats = get_stats_from_webhook(webhook_receiver, token=token_info.token)
+    if stats:
+        assert len(stats) == 1
+        assert stats[0]["memo"] == memo
+        _ = TokenAlertDetailGeneric(**stats[0])
```
Comment (gjcthinkst) [inline tests/integration/test_saml_token.py:54]:
```suggestion
        TokenAlertDetailGeneric(**stats[0])
```

nitpick

**Review by gjcthinkst — COMMENTED** [review]
Did we test this with both Azure and Okta?

What happens if users don't enter a redirect URL? Does the user get redirected to a default page after clicking on the app in their SAML provider?

File: frontend/app.py [hunk 3]
```diff
@@ -1954,10 +1961,7 @@ def _(token_request_details: MySQLTokenRequest, canarydrop: Canarydrop):
 def _(
     token_request_details: CreditCardV2TokenRequest, canarydrop: Canarydrop
 ) -> CreditCardV2TokenResponse:
-    canarytoken = Canarytoken()
-    canarydrop.canarytoken = canarytoken
-
-    (status, card) = credit_card_infra.create_card(canarytoken.value())
+    (status, card) = credit_card_infra.create_card(canarydrop.canarytoken.value())
```
[the remark this answers is not in the document]
Comment (wleightond) [reply frontend/app.py:1964]: Yup, the bug here is that if there's a webhook in use it sends a test webhook with the previous token value, and thereafter the token has a new value.

**Comment by wleightond** [comment]
@gjcthinkst 
> Did we test this with both Azure and Okta?

We did test with both (see also the docs [PR](https://github.com/thinkst/canarytokens-docs/pull/57)) :)

> What happens if users don't enter a redirect URL? Does the user get redirected to a default page after clicking on the app in their SAML provider?

The token is treated essentially as either a web bug or redirect token; so if they don't enter a redirect URL they get the pale blue dot or browser scanner page depending on setting, same as normal web bugs.

**Comment by gjcthinkst** [comment]
> @gjcthinkst
> 
> > Did we test this with both Azure and Okta?
> 
> We did test with both (see also the docs [PR](https://github.com/thinkst/canarytokens-docs/pull/57)) :)
> 
> > What happens if users don't enter a redirect URL? Does the user get redirected to a default page after clicking on the app in their SAML provider?
> 
> The token is treated essentially as either a web bug or redirect token; so if they don't enter a redirect URL they get the pale blue dot or browser scanner page depending on setting, same as normal web bugs.

Cool, thanks for clearing that up (and for making the changes)!

**Review by gjcthinkst — APPROVED** [review]

**Review by thomaswindell — APPROVED** [review]

**Pull request merged**